src/Og.php

```diff
@@ -183,6 +182,166 @@ public static function getEntityGroups(EntityInterface $entity, array $states =
+        continue;
+      }
+
+      // Compile a list of group target IDs.
+      $target_ids = array_map(function ($value) {
+        return $value['target_id'];
+      }, $entity->get($field->getName())->getValue());
+
+      // Query the database to get the actual list of groups. The target IDs may
+      // contain groups that no longer exist. Entityreference doesn't clean up
+      // orphaned target IDs.
+      $entity_type = \Drupal::entityTypeManager()->getDefinition($target_type);
+      $query = \Drupal::entityQuery($target_type)
+        ->condition($entity_type->getKey('id'), $target_ids, 'IN');
+
+      // Optionally filter by group bundle.
+      if (!empty($group_bundle)) {
+        $query->condition($entity_type->getKey('bundle'), $group_bundle);
+      }
+
+      $group_ids = NestedArray::mergeDeep($group_ids, [$target_type => $query->execute()]);
+    }
+
+    return $group_ids;
+  }
+
+  /**
+   * Returns all groups that are associated with the given group content entity.
+   *
+   * Do not use this to retrieve group memberships for a user entity. Use
+   * Og::GetEntityGroups() instead.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The group content entity for which to return the groups.
+   * @param string $group_type_id
+   *   Filter results to only include groups of this entity type.
+   * @param string $group_bundle
+   *   Filter results to only include groups of this bundle.
+   *
+   * @return \Drupal\Core\Entity\EntityInterface[]
+   *   An array of group entities.
+   */
+  public static function getGroups(EntityInterface $entity, $group_type_id = NULL, $group_bundle = NULL) {
+    $groups = [];
+
+    foreach (static::getGroupIds($entity, $group_type_id, $group_bundle) as $entity_type => $entity_ids) {
+      $entities = \Drupal::entityTypeManager()->getStorage($entity_type)->loadMultiple($entity_ids);
+      $groups = array_merge($groups, array_values($entities));
+    }
+
+    return $groups;
+  }
+
+  /**
+   * Returns the number of groups associated with a given group content entity.
+   *
+   * Do not use this to retrieve the group membership count for a user entity.
+   * Use count(Og::GetEntityGroups()) instead.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The group content entity for which to count the associated groups.
+   * @param string $group_type_id
+   *   Only count groups of this entity type.
+   * @param string $group_bundle
+   *   Only count groups of this bundle.
+   *
+   * @return int
+   *   The number of associated groups.
+   */
+  public static function getGroupCount(EntityInterface $entity, $group_type_id = NULL, $group_bundle = NULL) {
+    return array_reduce(static::getGroupIds($entity, $group_type_id, $group_bundle), function ($carry, $item) {
+      return $carry + count($item);
+    }, 0);
+  }
+
+  /**
+   * Returns all the group content associated with a given group entity.
```

Comment:
I had noticed the performance problem during testing and this is no longer returning group content entities, only group content IDs. I have forgotten to update the method name though. This should be called `::getGroupContentIds()` and the documentation should be updated accordingly.

I just tested it with a group containing 10000 group content items and it scales pretty well. The method takes 0.034 seconds on the first run, and 0.013 seconds on the second run (I guess because of the MySql query cache). 

```
$ drush php
Psy Shell v0.7.2 (PHP 7.0.5 — cli) by Justin Hileman
>>> $start = microtime(TRUE); echo count(\Drupal\og\Og::getGroupContent(\Drupal\block_content\Entity\BlockContent::load(5))['node']) . "\n" . (microtime(TRUE) - $start);
10000
0.034804105758667⏎
=> null
>>> $start = microtime(TRUE); echo count(\Drupal\og\Og::getGroupContent(\Drupal\block_content\Entity\BlockContent::load(5))['node']) . "\n" . (microtime(TRUE) - $start);
10000
0.012892007827759⏎
=> null
```

I guess this is acceptable performance :)